penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2131 @@
+  // Compute layering: TODO(@wodeni)
+  return { tag: "Just", contents: [] };
+};
+
+const computeShapeOrdering = (tr: Translation): string[] => {
+  const layeringExprs = findLayeringExprs(tr);
+  // Returns list of layering specifications [below, above]
+  const partialOrderings: [string, string][] = layeringExprs.map((e: Expr): [string, string] => findNames(e, tr));
+
+  const allGPINames: string[] = findShapeNames(tr)
+    .map((e: [string, Field]): string => getShapeName(e[0], e[1]));
+  const shapeOrdering = topSortLayering(allGPINames, partialOrderings);
+
+  // TODO: Errors for labeling
+  if (shapeOrdering.tag === "Nothing") {
+    throw Error("no shape ordering possible from layering");
+  }
+
+  return shapeOrdering.contents;
+};
+
+//#endregion
+
+// ---- MAIN FUNCTION
+
+// COMBAK: Add optConfig as param?
+const genOptProblemAndState = (trans: Translation): State => {
+
+  const varyingPaths = findVarying(trans);
+  // NOTE: the properties in uninitializedPaths are NOT floats. Floats are included in varyingPaths already
+  const uninitializedPaths = findUninitialized(trans);
+  const shapePathList: [string, string][] = findShapeNames(trans);
+  const shapePaths = shapePathList.map(mkPath);
+
+  // COMBAK: Use pseudorandomness
+  // sample varying fieldsr
+  const transInitFields = initFields(varyingPaths, trans);
+  // sample varying vals and instantiate all the non - float base properties of every GPI in the translation
+  const transInit = initShapes(transInitFields, shapePathList);
+
+  const shapeProperties = findShapesProperties(transInit);
+  const [objfnsDecl, constrfnsDecl] = findUserAppliedFns(transInit);
+  const [objfnsDefault, constrfnsDefault] = findDefaultFns(transInit);
+  const [objFns, constrFns] = [objfnsDecl.concat(objfnsDefault), constrfnsDecl.concat(constrfnsDefault)];
+
+  const [initialGPIs, transEvaled] = [[], transInit];
+  const initVaryingState = lookupNumericPaths(varyingPaths, transEvaled);
+  const pendingPaths = findPending(transInit);
+  const shapeOrdering = computeShapeOrdering(transInit); // deal with layering
+
+  const initState = {
+    shapes: initialGPIs, // These start out empty because they are initialized in the frontend via `evalShapes` in the Evaluator
+    shapePaths,
+    shapeProperties,
+    shapeOrdering,
+
+    translation: transInit, // This is the result of the data processing
+    originalTranslation: clone(trans),
+
+    varyingPaths,
+    varyingValues: initVaryingState,
+
+    uninitializedPaths,
+    pendingPaths,
+
+    objFns,
+    constrFns,
+
+    // `params` are initialized properly by optimization; the only thing it needs is the weight (for the objective function synthesis)
+    params: {
+      optStatus: { tag: "NewIter" },
+      weight: initConstraintWeight,
+      lbfgsInfo: defaultLbfgsParams,
+      UOround: -1,
+      EPround: -1,
+    } as unknown as Params,
```

Comment:
Minor: for now, you can use something like `const rng: prng = seedrandom(json.rng, { global: true });` for the generator. We can come back to this later.